### PR TITLE
fix(core): resolve links at boundary positions in editLink and deleteLink

### LINKED ISSUE #3073: Off-By-One `RangeError` Crash and Text Duplication in `StyleManager.editLink` and `deleteLink`
### What’s broken?

In `@blocknote/core`, `StyleManager.editLink` and `StyleManager.deleteLink` perform an unconditional `position + 1` lookup when locating the link mark at the current cursor position:

Inside [`packages/core/src/editor/managers/StyleManager.ts:220-260`](https://github.com/TypeCellOS/BlockNote/blob/main/packages/core/src/editor/managers/StyleManager.ts#L220-L260):

```
ts
public editLink(
  url: string,
  text: string,
  position = this.editor.transact((tr) => tr.selection.anchor),
) {
  this.editor.transact((tr) => {
    const linkData = this.getLinkMarkAtPos(position + 1);
    const { from, to } = linkData || {
      from: tr.selection.from,
      to: tr.selection.to,
    };
```
  

and in deleteLink: 
```
public deleteLink(
  position = this.editor.transact((tr) => tr.selection.anchor),
) {
  this.editor.transact((tr) => {
    const linkData = this.getLinkMarkAtPos(position + 1);
    const { from, to } = linkData || {
      from: tr.selection.from,
      to: tr.selection.to,
    };
    // 
```

This causes two major issues:

1. Unhandled RangeError Crash: If the link is positioned at the very end of the document (position === tr.doc.content.size), position + 1 resolves beyond document bounds. Inside getLinkMarkAtPos, tr.doc.resolve(pos) throws an uncaught RangeError: Position out of range, causing the editor to crash.
2. Text Duplication & Delete Failure: When the cursor caret is resting at the end of a link (position === link.to), position + 1 queries the character after the link. Since that node lacks the link mark, getLinkMarkAtPos returns undefined and falls back to { from: tr.selection.from, to: tr.selection.to }. Because the selection is a collapsed caret (from === to), editLink inserts the new text beside the old text without replacing it (resulting in OriginalTextEditedText duplication), and deleteLink attempts tr.removeMark(from, to, link) over an empty 0-length range, completely failing to remove the link.

### What did you expect to happen?

1. editLink and deleteLink should never throw an out-of-range error when the cursor is at the end of a document.
2. When the cursor is positioned at the boundary or end of a link (position === link.to), editLink should successfully replace the existing link text rather than duplicating it, and deleteLink should cleanly remove the link mark.

### Steps to reproduce

**Scenario A: RangeError Crash**

1. Create a document where a link is the last element: [GitHub](https://github.com)
2. Place the cursor at the end of the document (position === editor.prosemirrorState.doc.content.size).
3. Trigger editor.editLink("https://github.com", "GitHub Homepage") or editor.deleteLink().
4. Observed: Uncaught RangeError: Position <pos> out of range thrown from tr.doc.resolve(pos).

**Scenario B: Text Duplication / Failed Deletion**

1. Type a link: [Google](https://google.com) and place the cursor at the end of the text (position === link.to).
2. Call editor.editLink("https://google.com", "Google Search").
3. Observed: The text becomes GoogleGoogle Search instead of Google Search.
4. Call editor.deleteLink() with the cursor at the same position.
5. Observed: The link remains active and is not deleted.

### BlockNote version

Version: 0.54.0 (and main branch) Package: @blocknote/core

### Environment

OS: Any (Windows / macOS / Linux) Browsers: Chrome, Firefox, Safari, Edge Frameworks: Vanilla JS, React, Vue

### Additional context

**Root Cause**

getLinkMarkAtPos expects a valid in-bounds position. Unconditionally adding + 1 breaks on right-boundary cursor positions and document ends.

**Proposed Fix**

1. Guard getLinkMarkAtPos against positions exceeding tr.doc.content.size.
2. Inspect position directly, falling back to position - 1 if the cursor is at the trailing boundary of the link:

```
--- a/packages/core/src/editor/managers/StyleManager.ts
+++ b/packages/core/src/editor/managers/StyleManager.ts
@@ -154,6 +154,10 @@ export class StyleManager {
     return this.editor.transact((tr) => {
+      const clampedPos = Math.min(Math.max(0, pos), tr.doc.content.size);
+      const resolvedPos = tr.doc.resolve(clampedPos);
       const linkMark = resolvedPos
         .marks()
         .find((mark) => mark.type.name === "link");
@@ -224,7 +228,8 @@ export class StyleManager {
     this.editor.transact((tr) => {
-      const linkData = this.getLinkMarkAtPos(position + 1);
+      const linkData =
+        this.getLinkMarkAtPos(position) ||
+        (position > 0 ? this.getLinkMarkAtPos(position - 1) : undefined);
       const { from, to } = linkData || {
         from: tr.selection.from,
         to: tr.selection.to,
       };
@@ -248,7 +253,8 @@ export class StyleManager {
     this.editor.transact((tr) => {
-      const linkData = this.getLinkMarkAtPos(position + 1);
+      const linkData =
+        this.getLinkMarkAtPos(position) ||
+        (position > 0 ? this.getLinkMarkAtPos(position - 1) : undefined);
       const { from, to } = linkData || {
         from: tr.selection.from,
         to: tr.selection.to,
```

### Contribution

- [ ] I'd be interested in contributing a fix for this issue

### Sponsor

- [ ] I'm a [sponsor](https://www.blocknotejs.org/pricing) and would appreciate if you could look into this sooner than later 💖

### PR DESCRIPTION
**Pull request by Ishkirat-Singh**
# Summary

Fixes #3073. `StyleManager.editLink` and `deleteLink` looked up the link mark at `position + 1` unconditionally. At the end of the document that position does not exist and `tr.doc.resolve` throws a `RangeError`; at the end of a link (the usual place for the caret after clicking the link text) the lookup misses the link, falls back to the collapsed selection, and `editLink` inserts the new text next to the old link instead of replacing it.

## Rationale

The link mark is not inclusive, so a resolved position at either boundary does not carry the mark. Probing one position inside the link on both sides, and only accepting a link whose range actually contains the caret, finds the link at its start, inside it and at its end, without changing behaviour for a caret that is not on a link at all.

## Changes

- `packages/core/src/editor/managers/StyleManager.ts`: new private `getLinkMarkAround(position)` that probes `position + 1`, `position` and `position - 1`, skips positions outside the document, and returns the first link whose `[from, to]` range contains `position`. `editLink` and `deleteLink` use it instead of `getLinkMarkAtPos(position + 1)`.

## Impact

No public API change. Callers that pass a position outside the document no longer crash, and callers with the caret at a link boundary now edit or delete that link, which is the documented intent of both methods.

## Testing

New `packages/core/src/editor/managers/StyleManager.test.ts` (jsdom): edit and delete with the caret at the end of the link, edit with the caret at the start, and no throw for a position at the very end of the document. Ran it with `vp test --run` together with `vp lint src`.

## Checklist

- [x] Code follows the project's coding standards.
- [x] Unit tests covering the new feature have been added.
- [x] All existing tests pass.
- [ ] The documentation has been updated to reflect the new feature (no docs change needed).

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved link editing and deletion at the start and end boundaries of linked text.
  - Prevented errors when link actions are triggered at the end of a document.
  - Ensured link deletion correctly converts linked text to plain text in boundary cases.

- **Tests**
  - Added coverage for link editing and deletion at document and link boundaries.

### CODE REVIEW

**Comment by YousefED** [comment]
related https://github.com/TypeCellOS/BlockNote/pull/3058 (cc @nperez0111 )

**Comment by nperez0111** [comment]
Thanks for the attempt @Ishkirat-Singh, we decided to go with the other PR
